{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mImporting Data...\u001b[0m\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [45], line 70\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[39mprint\u001b[39m(adata)\n\u001b[1;32m     69\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m---> 70\u001b[0m     adata \u001b[39m=\u001b[39m sc\u001b[39m.\u001b[39;49mread(inputData)\n\u001b[1;32m     71\u001b[0m adata\u001b[39m.\u001b[39mvar_names_make_unique()\n\u001b[1;32m     73\u001b[0m geneMarkers \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\u001b[39m'\u001b[39m\u001b[39m./dataSaveOriginal/cellTypeMarkers.csv\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/scanpy/readwrite.py:112\u001b[0m, in \u001b[0;36mread\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, **kwargs)\u001b[0m\n\u001b[1;32m    110\u001b[0m filename \u001b[39m=\u001b[39m Path(filename)  \u001b[39m# allow passing strings\u001b[39;00m\n\u001b[1;32m    111\u001b[0m \u001b[39mif\u001b[39;00m is_valid_filename(filename):\n\u001b[0;32m--> 112\u001b[0m     \u001b[39mreturn\u001b[39;00m _read(\n\u001b[1;32m    113\u001b[0m         filename,\n\u001b[1;32m    114\u001b[0m         backed\u001b[39m=\u001b[39;49mbacked,\n\u001b[1;32m    115\u001b[0m         sheet\u001b[39m=\u001b[39;49msheet,\n\u001b[1;32m    116\u001b[0m         ext\u001b[39m=\u001b[39;49mext,\n\u001b[1;32m    117\u001b[0m         delimiter\u001b[39m=\u001b[39;49mdelimiter,\n\u001b[1;32m    118\u001b[0m         first_column_names\u001b[39m=\u001b[39;49mfirst_column_names,\n\u001b[1;32m    119\u001b[0m         backup_url\u001b[39m=\u001b[39;49mbackup_url,\n\u001b[1;32m    120\u001b[0m         cache\u001b[39m=\u001b[39;49mcache,\n\u001b[1;32m    121\u001b[0m         cache_compression\u001b[39m=\u001b[39;49mcache_compression,\n\u001b[1;32m    122\u001b[0m         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    123\u001b[0m     )\n\u001b[1;32m    124\u001b[0m \u001b[39m# generate filename and read to dict\u001b[39;00m\n\u001b[1;32m    125\u001b[0m filekey \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(filename)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/scanpy/readwrite.py:722\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, cache_compression, suppress_cache_warning, **kwargs)\u001b[0m\n\u001b[1;32m    720\u001b[0m \u001b[39mif\u001b[39;00m ext \u001b[39min\u001b[39;00m {\u001b[39m'\u001b[39m\u001b[39mh5\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mh5ad\u001b[39m\u001b[39m'\u001b[39m}:\n\u001b[1;32m    721\u001b[0m     \u001b[39mif\u001b[39;00m sheet \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 722\u001b[0m         \u001b[39mreturn\u001b[39;00m read_h5ad(filename, backed\u001b[39m=\u001b[39;49mbacked)\n\u001b[1;32m    723\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    724\u001b[0m         logg\u001b[39m.\u001b[39mdebug(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mreading sheet \u001b[39m\u001b[39m{\u001b[39;00msheet\u001b[39m}\u001b[39;00m\u001b[39m from file \u001b[39m\u001b[39m{\u001b[39;00mfilename\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/h5ad.py:236\u001b[0m, in \u001b[0;36mread_h5ad\u001b[0;34m(filename, backed, as_sparse, as_sparse_fmt, chunk_size)\u001b[0m\n\u001b[1;32m    233\u001b[0m     \u001b[39massert\u001b[39;00m \u001b[39mFalse\u001b[39;00m, \u001b[39m\"\u001b[39m\u001b[39munexpected raw format\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    234\u001b[0m \u001b[39melif\u001b[39;00m k \u001b[39min\u001b[39;00m {\u001b[39m\"\u001b[39m\u001b[39mobs\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mvar\u001b[39m\u001b[39m\"\u001b[39m}:\n\u001b[1;32m    235\u001b[0m     \u001b[39m# Backwards compat\u001b[39;00m\n\u001b[0;32m--> 236\u001b[0m     d[k] \u001b[39m=\u001b[39m read_dataframe(f[k])\n\u001b[1;32m    237\u001b[0m \u001b[39melse\u001b[39;00m:  \u001b[39m# Base case\u001b[39;00m\n\u001b[1;32m    238\u001b[0m     d[k] \u001b[39m=\u001b[39m read_elem(f[k])\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/h5ad.py:301\u001b[0m, in \u001b[0;36mread_dataframe\u001b[0;34m(group)\u001b[0m\n\u001b[1;32m    299\u001b[0m     \u001b[39mreturn\u001b[39;00m read_dataframe_legacy(group)\n\u001b[1;32m    300\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 301\u001b[0m     \u001b[39mreturn\u001b[39;00m read_elem(group)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/registry.py:183\u001b[0m, in \u001b[0;36mread_elem\u001b[0;34m(elem, modifiers)\u001b[0m\n\u001b[1;32m    178\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mread_elem\u001b[39m(\n\u001b[1;32m    179\u001b[0m     elem: Union[H5Array, H5Group, ZarrGroup, ZarrArray],\n\u001b[1;32m    180\u001b[0m     modifiers: \u001b[39mfrozenset\u001b[39m(\u001b[39mstr\u001b[39m) \u001b[39m=\u001b[39m \u001b[39mfrozenset\u001b[39m(),\n\u001b[1;32m    181\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Any:\n\u001b[1;32m    182\u001b[0m     \u001b[39m\"\"\"Read an element from an on disk store.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 183\u001b[0m     \u001b[39mreturn\u001b[39;00m _REGISTRY\u001b[39m.\u001b[39;49mget_reader(\u001b[39mtype\u001b[39;49m(elem), get_spec(elem), \u001b[39mfrozenset\u001b[39;49m(modifiers))(elem)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:564\u001b[0m, in \u001b[0;36mread_dataframe_0_1_0\u001b[0;34m(elem)\u001b[0m\n\u001b[1;32m    561\u001b[0m columns \u001b[39m=\u001b[39m _read_attr(elem\u001b[39m.\u001b[39mattrs, \u001b[39m\"\u001b[39m\u001b[39mcolumn-order\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    562\u001b[0m idx_key \u001b[39m=\u001b[39m _read_attr(elem\u001b[39m.\u001b[39mattrs, \u001b[39m\"\u001b[39m\u001b[39m_index\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    563\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(\n\u001b[0;32m--> 564\u001b[0m     {k: read_series(elem[k]) \u001b[39mfor\u001b[39;00m k \u001b[39min\u001b[39;00m columns},\n\u001b[1;32m    565\u001b[0m     index\u001b[39m=\u001b[39mread_series(elem[idx_key]),\n\u001b[1;32m    566\u001b[0m     columns\u001b[39m=\u001b[39m\u001b[39mlist\u001b[39m(columns),\n\u001b[1;32m    567\u001b[0m )\n\u001b[1;32m    568\u001b[0m \u001b[39mif\u001b[39;00m idx_key \u001b[39m!=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m_index\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m    569\u001b[0m     df\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39mname \u001b[39m=\u001b[39m idx_key\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:564\u001b[0m, in \u001b[0;36m<dictcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    561\u001b[0m columns \u001b[39m=\u001b[39m _read_attr(elem\u001b[39m.\u001b[39mattrs, \u001b[39m\"\u001b[39m\u001b[39mcolumn-order\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    562\u001b[0m idx_key \u001b[39m=\u001b[39m _read_attr(elem\u001b[39m.\u001b[39mattrs, \u001b[39m\"\u001b[39m\u001b[39m_index\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    563\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(\n\u001b[0;32m--> 564\u001b[0m     {k: read_series(elem[k]) \u001b[39mfor\u001b[39;00m k \u001b[39min\u001b[39;00m columns},\n\u001b[1;32m    565\u001b[0m     index\u001b[39m=\u001b[39mread_series(elem[idx_key]),\n\u001b[1;32m    566\u001b[0m     columns\u001b[39m=\u001b[39m\u001b[39mlist\u001b[39m(columns),\n\u001b[1;32m    567\u001b[0m )\n\u001b[1;32m    568\u001b[0m \u001b[39mif\u001b[39;00m idx_key \u001b[39m!=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m_index\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m    569\u001b[0m     df\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39mname \u001b[39m=\u001b[39m idx_key\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590\u001b[0m, in \u001b[0;36mread_series\u001b[0;34m(dataset)\u001b[0m\n\u001b[1;32m    586\u001b[0m     \u001b[39mreturn\u001b[39;00m pd\u001b[39m.\u001b[39mCategorical\u001b[39m.\u001b[39mfrom_codes(\n\u001b[1;32m    587\u001b[0m         read_elem(dataset), categories, ordered\u001b[39m=\u001b[39mordered\n\u001b[1;32m    588\u001b[0m     )\n\u001b[1;32m    589\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 590\u001b[0m     \u001b[39mreturn\u001b[39;00m read_elem(dataset)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/registry.py:183\u001b[0m, in \u001b[0;36mread_elem\u001b[0;34m(elem, modifiers)\u001b[0m\n\u001b[1;32m    178\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mread_elem\u001b[39m(\n\u001b[1;32m    179\u001b[0m     elem: Union[H5Array, H5Group, ZarrGroup, ZarrArray],\n\u001b[1;32m    180\u001b[0m     modifiers: \u001b[39mfrozenset\u001b[39m(\u001b[39mstr\u001b[39m) \u001b[39m=\u001b[39m \u001b[39mfrozenset\u001b[39m(),\n\u001b[1;32m    181\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Any:\n\u001b[1;32m    182\u001b[0m     \u001b[39m\"\"\"Read an element from an on disk store.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 183\u001b[0m     \u001b[39mreturn\u001b[39;00m _REGISTRY\u001b[39m.\u001b[39;49mget_reader(\u001b[39mtype\u001b[39;49m(elem), get_spec(elem), \u001b[39mfrozenset\u001b[39;49m(modifiers))(elem)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:94\u001b[0m, in \u001b[0;36mread_basic\u001b[0;34m(elem)\u001b[0m\n\u001b[1;32m     92\u001b[0m     \u001b[39mreturn\u001b[39;00m {k: read_elem(v) \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m elem\u001b[39m.\u001b[39mitems()}\n\u001b[1;32m     93\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(elem, h5py\u001b[39m.\u001b[39mDataset):\n\u001b[0;32m---> 94\u001b[0m     \u001b[39mreturn\u001b[39;00m h5ad\u001b[39m.\u001b[39;49mread_dataset(elem)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/utils.py:182\u001b[0m, in \u001b[0;36mreport_read_key_on_error.<locals>.func_wrapper\u001b[0;34m(elem, *args, **kwargs)\u001b[0m\n\u001b[1;32m    179\u001b[0m \u001b[39m@wraps\u001b[39m(func)\n\u001b[1;32m    180\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mfunc_wrapper\u001b[39m(elem, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    181\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 182\u001b[0m         \u001b[39mreturn\u001b[39;00m func(elem, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    183\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    184\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(e, AnnDataReadError):\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/anndata/_io/h5ad.py:310\u001b[0m, in \u001b[0;36mread_dataset\u001b[0;34m(dataset)\u001b[0m\n\u001b[1;32m    308\u001b[0m     \u001b[39mif\u001b[39;00m (string_dtype \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m) \u001b[39mand\u001b[39;00m (string_dtype\u001b[39m.\u001b[39mencoding \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mutf-8\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m    309\u001b[0m         dataset \u001b[39m=\u001b[39m dataset\u001b[39m.\u001b[39masstr()\n\u001b[0;32m--> 310\u001b[0m value \u001b[39m=\u001b[39m dataset[()]\n\u001b[1;32m    311\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mhasattr\u001b[39m(value, \u001b[39m\"\u001b[39m\u001b[39mdtype\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m    312\u001b[0m     \u001b[39mreturn\u001b[39;00m value\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/h5py/_hl/dataset.py:223\u001b[0m, in \u001b[0;36mAsStrWrapper.__getitem__\u001b[0;34m(self, args)\u001b[0m\n\u001b[1;32m    222\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__getitem__\u001b[39m(\u001b[39mself\u001b[39m, args):\n\u001b[0;32m--> 223\u001b[0m     bytes_arr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_dset[args]\n\u001b[1;32m    224\u001b[0m     \u001b[39m# numpy.char.decode() seems like the obvious thing to use. But it only\u001b[39;00m\n\u001b[1;32m    225\u001b[0m     \u001b[39m# accepts numpy string arrays, not object arrays of bytes (which we\u001b[39;00m\n\u001b[1;32m    226\u001b[0m     \u001b[39m# return from HDF5 variable-length strings). And the numpy\u001b[39;00m\n\u001b[1;32m    227\u001b[0m     \u001b[39m# implementation is not faster than doing it with a loop; in fact, by\u001b[39;00m\n\u001b[1;32m    228\u001b[0m     \u001b[39m# not converting the result to a numpy unicode array, the\u001b[39;00m\n\u001b[1;32m    229\u001b[0m     \u001b[39m# naive way can be faster! (Comparing with numpy 1.18.4, June 2020)\u001b[39;00m\n\u001b[1;32m    230\u001b[0m     \u001b[39mif\u001b[39;00m numpy\u001b[39m.\u001b[39misscalar(bytes_arr):\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/h5py/_hl/dataset.py:814\u001b[0m, in \u001b[0;36mDataset.__getitem__\u001b[0;34m(self, args, new_dtype)\u001b[0m\n\u001b[1;32m    809\u001b[0m     \u001b[39mreturn\u001b[39;00m arr\n\u001b[1;32m    811\u001b[0m \u001b[39m# === Everything else ===================\u001b[39;00m\n\u001b[1;32m    812\u001b[0m \n\u001b[1;32m    813\u001b[0m \u001b[39m# Perform the dataspace selection.\u001b[39;00m\n\u001b[0;32m--> 814\u001b[0m selection \u001b[39m=\u001b[39m sel\u001b[39m.\u001b[39;49mselect(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mshape, args, dataset\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m)\n\u001b[1;32m    816\u001b[0m \u001b[39mif\u001b[39;00m selection\u001b[39m.\u001b[39mnselect \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    817\u001b[0m     \u001b[39mreturn\u001b[39;00m numpy\u001b[39m.\u001b[39mzeros(selection\u001b[39m.\u001b[39marray_shape, dtype\u001b[39m=\u001b[39mnew_dtype)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/h5py/_hl/selections.py:82\u001b[0m, in \u001b[0;36mselect\u001b[0;34m(shape, args, dataset)\u001b[0m\n\u001b[1;32m     79\u001b[0m     space \u001b[39m=\u001b[39m h5s\u001b[39m.\u001b[39mcreate_simple(shape)\n\u001b[1;32m     80\u001b[0m     selector \u001b[39m=\u001b[39m _selector\u001b[39m.\u001b[39mSelector(space)\n\u001b[0;32m---> 82\u001b[0m \u001b[39mreturn\u001b[39;00m selector\u001b[39m.\u001b[39;49mmake_selection(args)\n",
      "File \u001b[0;32mh5py/_selector.pyx:295\u001b[0m, in \u001b[0;36mh5py._selector.Selector.make_selection\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.10/site-packages/h5py/_hl/selections.py:227\u001b[0m, in \u001b[0;36mSimpleSelection.__init__\u001b[0;34m(self, shape, spaceid, hyperslab)\u001b[0m\n\u001b[1;32m    224\u001b[0m     scalar \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_sel[\u001b[39m3\u001b[39m]\n\u001b[1;32m    225\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mtuple\u001b[39m(x \u001b[39mfor\u001b[39;00m x, s \u001b[39min\u001b[39;00m \u001b[39mzip\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmshape, scalar) \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m s)\n\u001b[0;32m--> 227\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, shape, spaceid\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, hyperslab\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    228\u001b[0m     \u001b[39msuper\u001b[39m()\u001b[39m.\u001b[39m\u001b[39m__init__\u001b[39m(shape, spaceid)\n\u001b[1;32m    229\u001b[0m     \u001b[39mif\u001b[39;00m hyperslab \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from itertools import groupby\n",
    "from json import load\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import sys\n",
    "import tests.loadScanpy as loadScanpy\n",
    "import classifyClusters as classify\n",
    "import os\n",
    "from scipy.sparse import csr_matrix\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import seaborn as sns\n",
    "import diffxpy.api as de\n",
    "from diffexpr.diffexpr.py_deseq import py_DESeq2\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "# Global Settings \n",
    "####################################################################################################\n",
    "np.set_printoptions(threshold=sys.maxsize)\n",
    "pd.options.display.max_columns = sys.maxsize\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "\n",
    "inputData = \"./dataSaveOriginal/rawDataset5000.h5ad\" # ./dataSaveOriginal/rawDataset5000.h5ad\n",
    "results_file = './output/savedData.h5ad'\n",
    "outputDirectory = \"./outputPDFs/\"\n",
    "\n",
    "# Arguments Settings\n",
    "\n",
    "sysArgs = sys.argv\n",
    "\n",
    "showPlots = False\n",
    "if \"-showPlots\" in sysArgs:\n",
    "    showPlots = True\n",
    "else:\n",
    "    showPlots = False\n",
    "\n",
    "arg = \"\"\n",
    "for word in sysArgs:\n",
    "    if word != \"main.py\" and word != \"-showPlots\":\n",
    "        arg = arg + word + \" \"\n",
    "        # if \"-\" not in word:\n",
    "        #     OUTPUT_DIR = outputDirectory+word+\"/\"\n",
    "        #     CHECK_FOLDER = os.path.isdir(OUTPUT_DIR)\n",
    "        #     if not CHECK_FOLDER:\n",
    "        #         os.makedirs(OUTPUT_DIR)\n",
    "        #     outputDirectory = OUTPUT_DIR\n",
    "class bcolors:\n",
    "    HEADER = '\\033[95m'\n",
    "    OKBLUE = '\\033[94m'\n",
    "    OKCYAN = '\\033[96m'\n",
    "    OKGREEN = '\\033[92m'\n",
    "    WARNING = '\\033[93m'\n",
    "    FAIL = '\\033[91m'\n",
    "    ENDC = '\\033[0m'\n",
    "    BOLD = '\\033[1m'\n",
    "    UNDERLINE = '\\033[4m'    \n",
    "\n",
    "####################################################################################################\n",
    "# Import Data \n",
    "####################################################################################################\n",
    "print(bcolors.FAIL + \"Importing Data...\" + bcolors.ENDC)\n",
    "if \"-loadGenomics\" in sysArgs:\n",
    "    adata = loadScanpy.readData()\n",
    "    print(adata)\n",
    "else:\n",
    "    adata = sc.read(inputData)\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "geneMarkers = pd.read_csv('./dataSaveOriginal/cellTypeMarkers.csv')\n",
    "geneNames = geneMarkers[\"gene\"].tolist()\n",
    "\n",
    "genesHSA21 = pd.read_csv(\"./dataSaveOriginal/HSA21genes.csv\")\n",
    "genesHSA21List = []\n",
    "\n",
    "for gene in genesHSA21[\"hgnc_symbol\"].tolist():\n",
    "    if type(gene) == str:\n",
    "        genesHSA21List.append(gene)\n",
    "\n",
    "# ####################################################################################################\n",
    "# # Pre-Processing\n",
    "# ####################################################################################################\n",
    "adataCON_DS2U =  adata[adata.obs[\"sample\"] == \"CON_DS2U\"]\n",
    "adataCON_H9 =  adata[adata.obs[\"sample\"] == \"CON_H9\"]\n",
    "adataCON_IMR =  adata[adata.obs[\"sample\"] == \"CON_IMR\"]\n",
    "adataCON_ihtc =  adata[adata.obs[\"sample\"] == \"CON_ihtc\"]\n",
    "adataDS_2DS3 =  adata[adata.obs[\"sample\"] == \"DS_2DS3\"]\n",
    "adataDS_DS1 =  adata[adata.obs[\"sample\"] == \"DS_DS1\"]\n",
    "adataDS_DSP =  adata[adata.obs[\"sample\"] == \"DS_DSP\"]\n",
    "adata = adataCON_DS2U.concatenate(adataCON_H9)\n",
    "adata = adata.concatenate(adataCON_IMR)\n",
    "adata = adata.concatenate(adataDS_2DS3)\n",
    "adata = adata.concatenate(adataDS_DSP)\n",
    "adata.raw = adata\n",
    "# adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "# riboURL = \"http://software.broadinstitute.org/gsea/msigdb/download_geneset.jsp?geneSetName=KEGG_RIBOSOME&fileType=txt\"\n",
    "# riboGenes = pd.read_table(riboURL, skiprows=2, header = None)\n",
    "# adata.var['ribo'] = adata.var_names.isin(riboGenes[0].values)\n",
    "# sc.pp.calculate_qc_metrics(adata, qc_vars=['mt', \"ribo\"], percent_top=None, log1p=False, inplace=True)\n",
    "# sc.pp.normalize_total(adata, target_sum=10000)\n",
    "# sc.pp.log1p(adata)\n",
    "# sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "# sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "# sc.pp.scale(adata, max_value=10)\n",
    "# sc.tl.pca(adata, svd_solver='arpack')\n",
    "# sc.pp.neighbors(adata, n_neighbors=15, n_pcs=40) # n_neighbors=15 is default\n",
    "# sc.tl.umap(adata)\n",
    "# sc.tl.leiden(adata, resolution=0.5)\n",
    "# sc.external.pp.bbknn(adata, batch_key='sample')\n",
    "\n",
    "# ####################################################################################################\n",
    "# # Differential Expression Using DESeq2\n",
    "# ####################################################################################################\n",
    "print(bcolors.FAIL + \"Performing Differential Expression using DESeq2...\" + bcolors.ENDC)\n",
    "def getColSample(string):\n",
    "    subset =  adata[adata.obs[\"sample\"] == string]\n",
    "    returnArray = [sum(x) for x in zip(*subset.X)]\n",
    "    print(string)\n",
    "    print(returnArray[0])\n",
    "    return returnArray[0].toarray()[0]\n",
    "\n",
    "rows = adata.var[\"features\"].values\n",
    "\n",
    "# CON_DS2U CON_H9 CON_IMR CON_ihtc DS_2DS3 DS_DS1 DS_DSP\n",
    "\n",
    "df = pd.DataFrame(rows, columns=['id'])\n",
    "df[\"N_1\"] = getColSample(\"CON_DS2U\")\n",
    "df[\"N_2\"] = getColSample(\"CON_H9\")\n",
    "df[\"N_3\"] = getColSample(\"CON_IMR\")\n",
    "df[\"S_1\"] = getColSample(\"DS_2DS3\")\n",
    "df[\"S_2\"] = getColSample(\"DS_DSP\")\n",
    "\n",
    "print(df.head(10))\n",
    "\n",
    "df.N_1 = df.N_1.astype(int)\n",
    "df.N_2 = df.N_2.astype(int)\n",
    "df.N_3 = df.N_3.astype(int)\n",
    "df.S_1 = df.S_1.astype(int)\n",
    "df.S_2 = df.S_2.astype(int)\n",
    "print(df.head(50))\n",
    "\n",
    "print(\"Setting Sample Dataframe...\")\n",
    "sample_df = pd.DataFrame({'samplename': df.columns}) \\\n",
    "        .query('samplename != \"id\"')\\\n",
    "        .assign(sample = lambda d: d.samplename.str.extract('([NS])_', expand=False)) \\\n",
    "        .assign(replicate = lambda d: d.samplename.str.extract('_([123])', expand=False)) \n",
    "sample_df.index = sample_df.samplename\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running DESeq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['sample', 'S', 'N']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHCHD2', 'ZNF558', 'COL22A1', 'LINC02028', 'OPRK1', 'AL353622.1', 'FAM242E', 'FAM135B', 'ANKRD18B', 'AC024022.1', 'PSTPIP1', 'LINC01210', 'AC007368.1', 'LINC02826', 'RSPO1', 'AC090348.1', 'DPP6', 'LHX5-AS1', 'AL122035.2', 'PM20D1', 'CEBPA-DT', 'LYNX1', 'AC011416.3', 'TCEAL5', 'TTC22', 'SKOR2', 'AC063960.2', 'CFAP73', 'MSR1', 'AL354863.1', 'HMCN1', 'VSX1', 'RSPO2', 'TERT', 'FREM2-AS1', 'AC092335.1', 'AC025508.1', 'LINC02058', 'BAALC-AS2', 'AC090809.1', 'TMEM236', 'LINC01954', 'IRX1', 'COL24A1', 'TMEM132D', 'ZCCHC24', 'AL161851.2', 'IKBKE', 'AC006974.2', 'LINC01121']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(\"Running DESeq2...\")\n",
    "dds = py_DESeq2(count_matrix = df,\n",
    "               design_matrix = sample_df,\n",
    "               design_formula = '~ replicate + sample',\n",
    "               gene_column = 'id') # <- telling DESeq2 this should be the gene ID column\n",
    "    \n",
    "dds.run_deseq() \n",
    "dds.get_deseq_result(contrast = ['sample','S','N'])\n",
    "res = dds.deseq_result \n",
    "\n",
    "res = res.sort_values(['log2FoldChange'], ascending = [True])\n",
    "\n",
    "res\n",
    "\n",
    "geneList = res[\"id\"].tolist()[0:50]\n",
    "print(geneList)\n",
    "\n",
    "# geneList = [\"LOXL3\", \"M1AP\"]\n",
    "\n",
    "plotDF = pd.DataFrame(columns=[\"CON\", \"DS\"]) # index = geneList\n",
    "i = 0\n",
    "for resultGene in geneList:\n",
    "    rowResult = df.loc[df['id'] == resultGene]\n",
    "    averageCON = float((rowResult[\"N_1\"]+rowResult[\"N_2\"]+rowResult[\"N_3\"])/3)\n",
    "    averageDS = float((rowResult[\"S_1\"]+rowResult[\"S_2\"])/2)\n",
    "    sumTotal = averageCON + averageDS\n",
    "    normalCON = float(averageCON/sumTotal)\n",
    "    normalDS = float(averageDS/sumTotal)\n",
    "    plotDF.loc[resultGene] = [averageCON, averageDS]\n",
    "    i +=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CHCHD2</th>\n",
       "      <td>1372.562229</td>\n",
       "      <td>-11.725420</td>\n",
       "      <td>1.504833</td>\n",
       "      <td>-7.791839</td>\n",
       "      <td>6.604108e-15</td>\n",
       "      <td>4.766846e-11</td>\n",
       "      <td>CHCHD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF558</th>\n",
       "      <td>39.923183</td>\n",
       "      <td>-8.398152</td>\n",
       "      <td>1.682090</td>\n",
       "      <td>-4.992690</td>\n",
       "      <td>5.954418e-07</td>\n",
       "      <td>4.775443e-04</td>\n",
       "      <td>ZNF558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL22A1</th>\n",
       "      <td>17.700327</td>\n",
       "      <td>-7.814851</td>\n",
       "      <td>4.787175</td>\n",
       "      <td>-1.632456</td>\n",
       "      <td>1.025835e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL22A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC02028</th>\n",
       "      <td>81.803829</td>\n",
       "      <td>-7.432582</td>\n",
       "      <td>4.344113</td>\n",
       "      <td>-1.710955</td>\n",
       "      <td>8.708943e-02</td>\n",
       "      <td>9.999587e-01</td>\n",
       "      <td>LINC02028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OPRK1</th>\n",
       "      <td>10.510254</td>\n",
       "      <td>-6.427551</td>\n",
       "      <td>2.142334</td>\n",
       "      <td>-3.000256</td>\n",
       "      <td>2.697529e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OPRK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC136616.3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC136616.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC136616.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC136616.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC141272.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC023491.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC007325.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE      stat        pvalue  \\\n",
       "CHCHD2      1372.562229      -11.725420  1.504833 -7.791839  6.604108e-15   \n",
       "ZNF558        39.923183       -8.398152  1.682090 -4.992690  5.954418e-07   \n",
       "COL22A1       17.700327       -7.814851  4.787175 -1.632456  1.025835e-01   \n",
       "LINC02028     81.803829       -7.432582  4.344113 -1.710955  8.708943e-02   \n",
       "OPRK1         10.510254       -6.427551  2.142334 -3.000256  2.697529e-03   \n",
       "...                 ...             ...       ...       ...           ...   \n",
       "AC136616.3     0.000000             NaN       NaN       NaN           NaN   \n",
       "AC136616.2     0.000000             NaN       NaN       NaN           NaN   \n",
       "AC141272.1     0.000000             NaN       NaN       NaN           NaN   \n",
       "AC023491.2     0.000000             NaN       NaN       NaN           NaN   \n",
       "AC007325.1     0.000000             NaN       NaN       NaN           NaN   \n",
       "\n",
       "                    padj          id  \n",
       "CHCHD2      4.766846e-11      CHCHD2  \n",
       "ZNF558      4.775443e-04      ZNF558  \n",
       "COL22A1              NaN     COL22A1  \n",
       "LINC02028   9.999587e-01   LINC02028  \n",
       "OPRK1                NaN       OPRK1  \n",
       "...                  ...         ...  \n",
       "AC136616.3           NaN  AC136616.3  \n",
       "AC136616.2           NaN  AC136616.2  \n",
       "AC141272.1           NaN  AC141272.1  \n",
       "AC023491.2           NaN  AC023491.2  \n",
       "AC007325.1           NaN  AC007325.1  \n",
       "\n",
       "[36601 rows x 7 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genesHSA21data = pd.DataFrame(columns=[\"baseMean\", \"log2FoldChange\", \"lfcSE\", \"stat\", \"pvalue\", \"padj\", \"id\"]) \n",
    "\n",
    "for index, row in genesHSA21data.iterrows():\n",
    "    rowGene = res.loc[res['id'] == row[\"id\"]]\n",
    "    print(rowGene)\n",
    "    genesHSA21data.loc[1] = rowGene\n",
    "\n",
    "\n",
    "genesHSA21data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "presMarkers = [\"MITF\", \"EMX2\", \"ZNF558\", \"KLF3\", \"ZNF589\", \"ZNF337\", \"ZNF16\", \"HMGA1\", \"ZIC2\", \"MECOM\", \"ESR2\", \"ZNF536\", \"RFXANK\", \"DLX1\", \"DLX2\", \"RORB\", \"OLIG2\", \"ZBTB20\", \"CREB3L2\", \"NPAS2\", \"ZNF491\", \"PBX3\", \"AFF2\", \"ZNF425\", \"ZHX3\", \"PKNOX1\", \"ARNT\", \"FOXN3\", \"GATAD2B\", \"ASH1L\"]\n",
    "\n",
    "with PdfPages(outputDirectory+'Differential Expression DESeq2 Test.pdf') as pdf:\n",
    "    dummyPlot = plt.plot([1, 2])\n",
    "    figureNum = plt.gcf().number\n",
    "\n",
    "    # sns.heatmap(plotDF, annot=True, cmap=sns.blend_palette([\"black\", \"green\"], as_cmap=True))\n",
    "    # plt.show()\n",
    "\n",
    "    genesHSA21List.remove('H2BS1')\n",
    "    genesHSA21List.remove('CFAP298-TCP10L')\n",
    "    genesHSA21List.remove('GET1-SH3BGR')\n",
    "    genesHSA21List.remove('SLX9')\n",
    "\n",
    "\n",
    "    sc.pl.heatmap(adata, geneList, groupby=\"group\", vmax=5, show=False)\n",
    "    sc.pl.heatmap(adata, geneList, groupby=\"sample\", vmax=5, show=False)\n",
    "    sc.pl.heatmap(adata, genesHSA21List, groupby=\"group\", vmax=5, show=False)\n",
    "    sc.pl.heatmap(adata, genesHSA21List, groupby=\"sample\", vmax=5, show=False)\n",
    "\n",
    "    for fig in range(figureNum+1,  plt.gcf().number+1):\n",
    "        pdf.savefig(figure=fig, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b0fa6594d8f4cbf19f97940f81e996739fb7646882a419484c72d19e05852a7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
